docs/src/main/asciidoc/_developer-guide/JPA_Reference_Guide.adoc

```diff
@@ -731,10 +664,89 @@ import org.hibernate.SessionFactory;
+----
+
+
+Applied Transformations to application classes compiled with Hibernate 5.1: ::
+*  In user type classes, change methods with parameter type org.hibernate.engine.spi.SessionImplementor to instead use type org.hibernate.engine.spi.SharedSessionContractImplementor, this is done for implementations of:
+**  org.hibernate.usertype.UserType.
+**  org.hibernate.usertype.CompositeUserType.
+**  org.hibernate.usertype.UserCollectionType.
+**  org.hibernate.usertype.UserVersionType.
+**  org.hibernate.type.Type.
+**  org.hibernate.type.SingleColumnType.
+**  org.hibernate.type.AbstractStandardBasicType.
+**  org.hibernate.type.ProcedureParameterExtractionAware.
+**  org.hibernate.type.ProcedureParameterNamedBinder.
+**  org.hibernate.type.VersionType.
+*  In user type classes, when calling hibernate.orm methods, cast SessionImplementor to SharedSessionContractImplementor.
+*  Change calls to org.hibernate.BasicQueryContract.getFlushMode() to instead call BasicQueryContract.getHibernateFlushMode().
+*  Change calls to org.hibernate.Session.getFlushMode() to instead call Session.getHibernateFlushMode().
+*  Change calls to org.hibernate.Query.getFirstResult(), to instead call Query.getHibernateFirstResult(), so null can be returned when the value is uninitialized (note that 0 is now returned instead of negative values).
+*  Change calls to org.hibernate.Query.getMaxResults(), to instead call Query.getHibernateMaxResults(), so that null will be returned when the value is uninitialized or org.hibernate.Query#setMaxResults() was called with a value <= 0.
+*  Change calls to org.hibernate.Query.setFirstResult(), to instead call Query.setHibernateFirstResult(), so that calls to set a value < 0 results in pagination starting with the 0th row as was done in Hibernate ORM 5.1.
+*  Change calls to org.hibernate.Query.setMaxResults(), to instead call Query.setHibernateMaxResults(), so that passed values <= 0 are treated the same as uninitialized.
+*  Change references to NEVER field in enum org.hibernate.FlushMode, to instead reference FlushMode.MANUAL.
+
+****
+TRACE + DEBUG level logging can be enabled via the "org.jboss.as.hibernate.transformer" logger category.
+Example output:
+[source, text]
+Hibernate51CompatibilityTransformer transformed application classes in 'deployment.org.jboss.as.test.compat.jpa.hibernate.transformer.VerifyHibernate51CompatibilityPropertyAndJDSEnabledTransformerTestCase.ear', class 'org/jboss/as/test/compat/jpa/hibernate/transformer/BitSetType' is calling method org/hibernate/type/AbstractSingleColumnStandardBasicType.replace, which must be changed to use SharedSessionContractImplementor as parameter.
+****
+
+
+
+[cols=",",options="header"]
+|=======================================================================
+|System Property |Purpose
+
+|Hibernate51CompatibilityTransformer=false| if true, the transformer will transform ORM 5.1.x native calls to, ORM 5.3 native calls, in every application deployment.
```

Comment:
I'm guessing that `Hibernate51CompatibilityTransformer=false` above mean that `false` is the default. It would be more clear to have a separate column, e.g. "Default", for the default value.

Reply:
great idea, thanks, adding a separate column.